### ISSUE TITLE
DI throws but renaming component class fixes it

### ISSUE THREAD

**Issue by zoechi** [dart-lang/angular#513]
```dart
class ProfileSummariesComponent {
  SearchPagerShared pager;
  ProfileSummariesComponent(this.pager);
}
```
throws 

```
EXCEPTION: No provider found for SearchPagerShared.
```

Renaming the component class to `XProfileSummariesComponent` fixes the problem.

To me this looks like a caching problem where somehow some invalid information is used.
But killing `pub serve`, deleting `.pub` in the project directory, and `~/.dartServer` didn't fix it.
Where else could it be cached?

**Comment by zoechi** [comment]
Seems it's not related to caching.
It seems just random changes to fix the exception.

**Comment by matanlurey** [comment]
This is in 3x, right?

**Comment by zoechi** [comment]
This is 
Angular 3.1.0
Dart VM version: 1.25.0-dev.4.0 (Thu Jun 29 02:23:46 2017) on "macos_x64"

**Comment by matanlurey** [comment]
I regret to inform you that the few folks with the most knowledge of barback/pub transformers are on vacation, so it's unlikely anyone will be able to help address this in the next week or so. 

/cc @natebosch @jakemac53 for when they are back and @kevmoo for triage.

**Comment by zoechi** [comment]
Is angular_components for Angular 4.0 near?
If it would be available I could at least try if 4.0 fixes the problem.

**Comment by matanlurey** [comment]
I imagine you could create a fork and `s/angular2/angular` and it would just work.

But I don't have any other information available re: timing.

**Comment by zoechi** [comment]
I thought about that. Thanks for confirming that you think this could work.

**Comment by matanlurey** [comment]
I have a confirmation internally it does, or at least did :)

We might be able to get out `angular_components`  w/ 4.0.0-alpha support relatively _soon_.

**Comment by natebosch** [comment]
To make sure I'm clear - the issue here is that failures are not consistent, that it's possible to get into a state where there are failures on code that otherwise should succeed?

Can you reliably repro this?

**Comment by zoechi** [comment]
@natebosch The project is currently in a status where it reliably fails.

While making arbitrary changes it started to work yesterday, but I have no reliable way to make it work now.
The changes that yesterday seemed to have "fixed" the problem don't help anymore. 

I tried to make a minimal reproduction of the exception in a new project, but wasn't able to reproduce. In this project similar code works just fine.
(The project is a customer project consisting of several Dart packages.)

The behavior seems similar to what I mentioned in #351 or https://github.com/dart-lang/angular/issues/348
where some errors seem to have been swallowed somewhere in `pub serve` or transformers which led to a similar weird behavior.

**Comment by matanlurey** [comment]
`angular_components` was released with 4.0.0 support. 

I'm marking as closed as we can't do anything else from our end without seeing the app.
